File: BudgetShopping.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;



class Result4 {

    /*
     * Complete the 'budgetShopping' function below.
     *
     * The function is expected to return an INTEGER.
     * The function accepts following parameters:
     *  1. INTEGER n
     *  2. INTEGER_ARRAY bundleQuantities
     *  3. INTEGER_ARRAY bundleCosts
     */

    public static int budgetShopping(int n, List<Integer> bundleQuantities, List<Integer> bundleCosts) {

        int m = bundleQuantities.size();
        List<Integer> maxQ = new ArrayList<Integer>();
        for(int i=0; i<n+1 ; i++)
            maxQ.add(0);

        for(int cost=0; cost<n+1; cost++){
            for(int j=0; j<m; j++){
                if( cost == 0 )
                    maxQ.set(cost, 0);
                else if(cost < bundleCosts.get(j))
                    maxQ.set(cost, (maxQ.get(cost-1) > maxQ.get(cost)) ? maxQ.get(cost-1) : maxQ.get(cost) );
                else
                    maxQ.set(cost,  maxQ.get(cost-1) > bundleQuantities.get(j)+maxQ.get(cost-bundleCosts.get(j))
                            ? ( maxQ.get(cost-1) > maxQ.get(cost) ? maxQ.get(cost-1) : maxQ.get(cost) )
                            : ( bundleQuantities.get(j)+maxQ.get(cost-bundleCosts.get(j)) > maxQ.get(cost) ? bundleQuantities.get(j)+maxQ.get(cost-bundleCosts.get(j)) : maxQ.get(cost) ));
            }
        }
        return maxQ.get(n);

    }

}

public class BudgetShopping {
    public static void main(String[] args) throws IOException {
        BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int n = Integer.parseInt(bufferedReader.readLine().trim());

        int bundleQuantitiesCount = Integer.parseInt(bufferedReader.readLine().trim());

        List<Integer> bundleQuantities = IntStream.range(0, bundleQuantitiesCount).mapToObj(i -> {
            try {
                return bufferedReader.readLine().replaceAll("\\s+$", "");
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        })
                .map(String::trim)
                .map(Integer::parseInt)
                .collect(toList());

        int bundleCostsCount = Integer.parseInt(bufferedReader.readLine().trim());

        List<Integer> bundleCosts = IntStream.range(0, bundleCostsCount).mapToObj(i -> {
            try {
                return bufferedReader.readLine().replaceAll("\\s+$", "");
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        })
                .map(String::trim)
                .map(Integer::parseInt)
                .collect(toList());

        int result = Result4.budgetShopping(n, bundleQuantities, bundleCosts);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedReader.close();
        bufferedWriter.close();
    }
}

File: Budget_Shopping.py
#!/bin/python

import math
import os
import random
import re
import sys



#
# Complete the 'budgetShopping' function below.
#
# The function is expected to return an INTEGER.
# The function accepts following parameters:
#  1. INTEGER n
#  2. INTEGER_ARRAY bundleQuantities
#  3. INTEGER_ARRAY bundleCosts
#

def budgetShopping(n, bundleQuantities, bundleCosts):

    m = len(bundleQuantities)
    q = [0]*(n+1)

    for i in range(n+1):
        for j in range(m):
            if i == 0:
                q[i] = 0
            elif i < bundleCosts[j]:
                q[i] = max(q[i-1], q[i])
            else:
                q[i] = max(q[i-1], bundleQuantities[j]+q[i-bundleCosts[j]], q[i])

    return q[n]


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(raw_input().strip())

    bundleQuantities_count = int(raw_input().strip())

    bundleQuantities = []

    for _ in xrange(bundleQuantities_count):
        bundleQuantities_item = int(raw_input().strip())
        bundleQuantities.append(bundleQuantities_item)

    bundleCosts_count = int(raw_input().strip())

    bundleCosts = []

    for _ in xrange(bundleCosts_count):
        bundleCosts_item = int(raw_input().strip())
        bundleCosts.append(bundleCosts_item)

    result = budgetShopping(n, bundleQuantities, bundleCosts)

    fptr.write(str(result) + '\n')

    fptr.close()
